catroid/src/main/java/org/catrobat/catroid/formulaeditor/datacontainer/SpriteDataBehaviour.java

```diff
@@ -0,0 +1,202 @@
+	protected abstract String getDataName(V data);
+	protected abstract void setDataName(V data, String name);
+	protected abstract String getKeyName(K key);
+	protected abstract void setValue(V dataToAdd, Object value);
+	protected abstract boolean isClone(K key);
+
+	V add(K key, String name) {
+		List<V> dataList = getOrCreate(key);
+		V dataToAdd = newInstance(name);
+		dataList.add(dataToAdd);
+		return dataToAdd;
+	}
+
+	V add(K key, String name, Object value) {
+		List<V> dataList = getOrCreate(key);
+		V dataToAdd = newInstance(name, value);
+		dataList.add(dataToAdd);
+		return dataToAdd;
+	}
+
+	V addIfNotExists(K key, String name) {
+		List<V> list = getDataMap().get(key);
+
+		if (list == null) {
+			list = new ArrayList<>();
+		}
+
+		for (V data : list) {
+			if (getDataName(data).equals(name)) {
+				return data;
+			}
+		}
+
+		V addedData = newInstance(name);
+		list.add(addedData);
+		getDataMap().put(key, list);
+
+		return addedData;
+	}
+
+	V addIfNotExists(K key, String name, Object value) {
+		V dataToAdd = addIfNotExists(key, name);
+		setValue(dataToAdd, value);
+		return dataToAdd;
+	}
+
+	V get(K key, String name) {
+		List<V> dataList = getOrCreate(key);
+		return find(dataList, name);
+	}
+
+	List<V> getOrCreate(K key) {
+		List<V> data = getDataMap().get(key);
+
+		if (data == null) {
+			data = new ArrayList<>();
+			getDataMap().put(key, data);
+		}
+		return data;
+	}
+
+	V find(List<V> dataList, String name) {
+		if (dataList == null) {
+			return null;
+		}
+		for (V data : dataList) {
+			if (getDataName(data).equals(name)) {
+				return data;
+			}
+		}
+		return null;
+	}
+
+	boolean exists(K key, V value) {
+		List<V> list = getDataMap().get(key);
+		return list != null && list.contains(value);
+	}
+
+	boolean exists(K key, String name) {
+		if (key == null) {
+			return false;
+		}
+		List<V> list = getDataMap().get(key);
+		if (list == null) {
+			return false;
+		}
+
+		for (V data : list) {
+			if (getDataName(data).equals(name)) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	boolean existsAny(List<K> keys, String name) {
+		for (K key : keys) {
+			if (exists(key, name)) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	boolean delete(K key, String name) {
+		if (key == null) {
+			return false;
+		}
+
+		List<V> dataList = getDataMap().get(key);
+		V dataToDelete = find(dataList, name);
+		if (dataToDelete != null) {
+			dataList.remove(dataToDelete);
+			return true;
+		}
+		return false;
+	}
+
+	Map<Sprite, List<V>> cloneForScene(Scene scene, Map<K, List<V>> dataMap) {
+		Map<Sprite, List<V>> clonedList = new HashMap<>();
+
+		for (Map.Entry<K, List<V>> entry : dataMap.entrySet()) {
+			List<V> newList = new ArrayList<>();
+			for (V data : entry.getValue()) {
+				newList.add(newInstance(getDataName(data)));
+			}
+
+			if (entry.getKey() == null) {
+				continue;
+			}
+
+			String keyName = getKeyName(entry.getKey());
+			Sprite clonedInstance = scene.getSpriteBySpriteName(keyName);
```

Comment:
`scene.getSpriteBySpriteName()` does NOT clone the instance, its simply a reference to the same object